### PR TITLE
fix(sessions): keep evicted subagent parents in the import window

### PR DESCRIPTION
**Pull request by carlotestor**
Replaces #7031 with a much smaller fix.

## Problem

Some Subagent Sessions render as top-level left-toolbar rows instead of children.

Lineage in `state.db` is correct. The issue is the import window: `read_importable_agent_session_rows()` applied the visible limit as a flat per-row recency slice (`projected[:limit]`). A subagent row is only renderable as a child when its parent row is in the same payload — the client's `_attachChildSessionsToSidebarRows` promotes a child to top level when the parent is missing.

An orchestrator parent is frozen while its children run, so it stops writing. Its still-streaming leaves win the recency race and the parent gets sliced out of the window, orphaning the leaves.

(#5305 doesn't cover this: it only suppresses when `_cross_surface_child_session`, which is set when `parent.source != child.source`. That marks `webui→subagent` but not `subagent→subagent`.)

## Fix

29 lines in `read_importable_agent_session_rows()`. After the recency slice, re-add subagent parents that are **already in `projected`** — the candidate query oversamples `limit * 8`, so these rows are in memory already.

- No extra SQL, no new helper module, no client changes.
- `CLI_VISIBLE_SESSION_LIMIT` stays 20.
- webui ancestors are deliberately **not** imported — that sidebar bucket already owns them.
- Terminates via the `have` set (cycle-safe), and only walks subagent→subagent edges.

Children are shown **as children**, not hidden.

## Why not #7031

#7031 (~304 production lines: extra SQL hops, re-projection with candidate rows, cycle/filter/hop-cap helpers) was correct but far more machinery than this bug needs. Closed in favour of this.

## Tests

`tests/test_subagent_parent_in_import_window.py` (82 lines), black-box through `read_importable_agent_session_rows`:

1. quiet subagent parent survives the slice; leaves keep `parent_session_id` + `relationship_type=child_session`
2. an unrelated older subagent row stays excluded (only ancestors are rescued)
3. a webui parent is not force-imported

Verified the first test fails without the production change (parent absent, leaves orphaned) and passes with it.

Supersedes #7031.

### CODE REVIEW

File: api/agent_sessions.py [hunk 1]
```diff
@@ -766,7 +766,35 @@ def read_importable_agent_session_rows(
...
+        while pending:
+            row = pending.pop()
+            if str(row.get('raw_source') or row.get('source') or '').strip().lower() != 'subagent':
+                continue
+            parent_id = row.get('parent_session_id')
+            if not parent_id or parent_id in have:
+                continue
+            parent = by_id.get(parent_id)
+            if parent is None:
+                continue
```
[the remark this answers is not in the document]
Comment (carlotestor) [reply api/agent_sessions.py:812]: You're right about the mechanism, and I reproduced the exact boundary rather than reasoning about it. With `limit=3` (so `candidate_limit = limit * 8 = 24`), a quiet parent plus N newer unrelated rows plus one hot child:

```
filler=22  total=24  orch_recovered=True    <- parent is candidate #24 of 24
filler=23  total=25  orch_recovered=False   <- parent is candidate #25, not fetched
filler=200 total=202 orch_recovered=False
```

So the failure is real and it lands precisely where you said: outside the oversample, `by_id.get(parent_id)` misses and the child renders top-level.

What I disagree with is that this should block the merge, because it is the *bound of the design*, not a regression this PR introduces. On `master` today that child is top-level in **every** case; after this PR it is top-level only when the parent is older than the 24 newest candidates. The change strictly narrows the broken set — it never widens it.

Closing the remaining gap means resolving arbitrary-age ancestors, which needs an unbounded per-row ancestor lookup (or a recursive CTE) on the hot sidebar path that runs on every `/api/sessions` poll. That is exactly the ~300-line approach in #7031, which was closed unmerged in favour of this one; re-adding it here would trade a rare cosmetic nesting miss for a query-cost regression on the interactive path.

So instead of hiding the bound, I documented and pinned it in 0cc073a0:

- **Docstring** — recovery is bounded by the `limit * 8` candidate set, an older ancestor stays unresolved, and widening it is a `candidate_limit` change rather than a change to this walk.
- **Inline comment** at the lines you flagged — same bound stated where the `by_id.get()` miss actually happens.
- **Tests** — you correctly noted the suite "does not exercise candidate-window exhaustion". Now it does, from both sides: `test_parent_inside_candidate_oversample_is_recovered` (filler=22, recovered) and `test_parent_beyond_candidate_oversample_stays_unresolved` (filler=23, not recovered). The second one pins the boundary so that if anyone later widens or narrows the oversample, it is a deliberate decision with a failing test, not a silent drift.

I also added two cases you didn't raise but which the walk should guarantee: `test_recovered_parent_is_returned_beyond_limit` (the return can exceed `limit` — I checked all four call sites in `api/models.py` and `api/gateway_watcher.py` and every one iterates, none assumes `len(rows) <= limit`) and `test_parent_cycle_does_not_hang` (a corrupt mutual `parent_session_id` link must not spin the `pending` loop — the `have` set already prevents it; the test locks that in).

7 tests pass, no behaviour change in this commit.

**Comment by carlotestor** [comment]
## Greptile review addressed — 0cc073a0

| # | Greptile point | Outcome |
|---|---|---|
| 1 | **P1 — Bounded candidates leave parents orphaned**: a selected child whose parent ranks below the newest `limit * 8` is absent from `projected`, so `by_id.get(parent_id)` skips it and the child is still promoted top-level | **Confirmed and reproduced, documented + pinned rather than "fixed"** — see [threaded reply](https://github.com/nesquena/hermes-webui/pull/7051#discussion_r3789169679). It is the bound of the design, not a regression this PR adds. |
| 2 | Summary: tests "do not exercise candidate-window exhaustion" | **Implemented** — two new tests cover both sides of the boundary. |

### On the P1

I measured the boundary instead of assuming it. With `limit=3` (`candidate_limit = 24`):

```
filler=22  parent is candidate #24 of 24  -> recovered
filler=23  parent is candidate #25 of 24  -> NOT recovered
```

Exactly as described. But the merge-blocking framing doesn't hold: on `master` that child is top-level in **100%** of cases; with this PR it is top-level only when the parent is older than the 24 newest candidates. The change strictly shrinks the broken set and never grows it.

Closing the residual gap requires an unbounded per-row ancestor lookup (or recursive CTE) on the sidebar path that runs on every `/api/sessions` poll — the ~300-line approach in #7031, closed unmerged in favour of this one. I'm not reintroducing a query-cost regression on the hot interactive path to fix a rare cosmetic nesting miss.

### What changed in 0cc073a0

- `api/agent_sessions.py` (+21, docs only, **no behaviour change**)
  - Docstring: `limit` bounds the *recency slice*, not the row count; recovery is bounded by the `limit * 8` oversample; widening it is a `candidate_limit` change, not a change to this walk.
  - Inline comment at the flagged lines, stating the bound where the `by_id.get()` miss actually occurs.
- `tests/test_subagent_parent_in_import_window.py` (+65)
  - `test_parent_inside_candidate_oversample_is_recovered` — filler=22, parent restored.
  - `test_parent_beyond_candidate_oversample_stays_unresolved` — filler=23, parent stays out. Pins the boundary so any future change to the oversample is deliberate and test-visible.
  - `test_recovered_parent_is_returned_beyond_limit` — the documented over-limit contract. All four call sites (`api/models.py` ×3, `api/gateway_watcher.py` ×1) iterate the result; none assumes `len(rows) <= limit`.
  - `test_parent_cycle_does_not_hang` — corrupt mutual `parent_session_id` must not spin the walk (the `have` set already prevents it; now locked in).

### Verification

```
$ pytest tests/test_subagent_parent_in_import_window.py -v
7 passed in 2.89s

$ pytest tests/test_issue3930_source_filter_pushdown.py \
         tests/test_issue3172_cron_session_limit.py \
         tests/test_issue3238_orphaned_cli_sidecar_prune.py
26 passed in 3.55s
```

Production diff across both commits stays at 29 changed lines in `agent_sessions.py`; everything added here is docstring, comment, and test. `CLI_VISIBLE_SESSION_LIMIT` is untouched.

**Comment by nesquena-hermes** [comment]
## 🔬 Exact-head gate — SAFE TO SHIP (queued for maintainer merge)

Thanks @carlotestor — clean, well-scoped fix for the frozen-orchestrator sidebar bug. `read_importable_agent_session_rows()` applied the visible-window `limit` as a flat per-row recency slice, so a quiet orchestrator whose leaves keep streaming lost the recency race, fell outside the window, and its subagent leaves got promoted to top-level sidebar rows. The re-add walks the already-projected candidate set (no extra query) to pull evicted subagent parents back in.

**Full authoritative gate:**
- **Codex adversarial: SAFE TO SHIP.** Verified: (1) the `while pending` loop **terminates** for both `a→b→a` cycles and self-parent lineage (the `have` guard prevents re-processing); (2) **no production caller assumes `len(result) <= limit`** — re-adding parents past the limit affects no pagination/cursor behavior; (3) source normalization + visibility preserve **subagent-only scope**, WebUI ancestors stay correctly excluded; (4) the added regression test **fails on the merge-base and passes on the PR** (non-vacuous), plus 10 focused projection/limit/attachment tests pass.
- **Full suite: 14,582 passed, 0 failures.**
- Honest scoping note (not a blocker): parents outside the `limit × 8` oversampled candidate window remain a pre-existing edge case this PR doesn't try to solve — it's not a regression introduced here.

Tight, additive, pure-backend, real regression test, near-zero risk. Queuing `gate-pass` for a maintainer merge nod.

*(Release-manager overnight gate — queued for maintainer merge, not autonomously tagged.)*

**Comment by carlotestor** [comment]
Fixed the one failing check (`test (3.13, 4)`) in 4becbb7c — all 23 checks are now green.

**Failing test**

```
tests/test_issue2513_custom_provider_remote_models.py::test_custom_provider_model_field_does_not_block_remote_catalog

>       assert "@custom:alpha-proxy:alpha/remote" in alpha_ids
E       AssertionError: assert '@custom:alpha-proxy:alpha/remote' in {'alpha/sticky'}
tests/test_issue2513_custom_provider_remote_models.py:96: AssertionError
------------------------------ Captured log call -------------------------------
WARNING  api.config:config.py:8355 live provider-catalog rebuild exceeded 4.0s budget — serving fallback, refreshing catalog out-of-band
```

**Root cause — pre-existing wall-clock flake, not a regression from this PR**

The captured warning is the tell. `get_available_models()` bounds a cold live provider-catalog rebuild with `_LIVE_REBUILD_BUDGET_SECONDS` (default 4s). This test never pinned that knob, so it raced the budget: on a starved runner the rebuild overruns, the degraded fallback catalog is served, and the `urlopen`-fetched `alpha/remote` is dropped — leaving only the config-declared sticky model `alpha/sticky`, exactly as asserted.

Evidence it is not from this branch:

- This PR touches only `api/agent_sessions.py` and `tests/test_subagent_parent_in_import_window.py`. It does not touch `api/config.py` or this test file.
- Both `api/config.py` and this test file are byte-identical to `origin/master` (md5 match; `diff` exit 0). Their last commits predate this branch.
- The same shard passed on 3.11 and 3.12 — same test set, different timing.
- Reproduced the exact CI assertion on **unmodified master** by shrinking the budget, and again at the stock 4s budget by simulating a slow rebuild — byte-identical output including `4.0s`, `config.py:8355` and `:96`.

**Change** — one `monkeypatch` line (plus comment) in `tests/test_issue2513_custom_provider_remote_models.py`, forcing the synchronous unbounded rebuild path:

```python
monkeypatch.setattr(config, "_LIVE_REBUILD_BUDGET_SECONDS", 0.0, raising=False)
```

`budget <= 0` selects the "legacy synchronous (unbounded) rebuild" branch (`api/config.py:8188-8189`), removing the timing dimension entirely. This matches existing precedent in `tests/test_issue2540_models_endpoint_error.py:20-24`, which pins the same knob for the same reason. No production code touched.

**A/B verification in the same worktree** — unpatched, at a budget hostile enough to force the fallback:

```
E   AssertionError: assert '@custom:alpha-proxy:alpha/remote' in {'alpha/sticky'}
WARNING  api.config:config.py:8355 live provider-catalog rebuild exceeded 0.0s budget — serving fallback
1 failed in 2.11s
```

Patched, same command: `1 passed in 2.09s` — and also passes at budget `0` and at the default. `ruff check` clean; this PR's own `tests/test_subagent_parent_in_import_window.py` still passes 7/7.

**CI after the push (`4becbb7c`)**

```
test (3.13, 4)   pass   3m1s
...
23 pass
```

**Comment by nesquena-hermes** [comment]
Shipped in experimental release **exp-v0.52.229** (batched with 5 other low-risk gate-passed fixes). Merged onto master via the batch release PR #7089 — your commit content is in master and is live on the experimental channel now.

Thanks @carlotestor for the fix.

🤖 Closed by the release-manager agent.
